numba_dpex/decorators.py

```diff
@@ -27,23 +28,27 @@ def kernel(signature=None, access_types=None, debug=None):
         * All array arguments passed to a kernel should be of the same type
           and have the same dtype.
     """
-    if signature is None:
+    if func_or_sig is None:
         return autojit(debug=debug, access_types=access_types)
-    elif not sigutils.is_signature(signature):
-        func = signature
+    elif not sigutils.is_signature(func_or_sig):
+        func = func_or_sig
         return autojit(debug=debug, access_types=access_types)(func)
     else:
-        return _kernel_jit(signature, debug, access_types)
+        return _kernel_jit(func_or_sig, debug, access_types)
```

Comment:
We need to add some example/test cases to show how to reach `_kernel_jit()`.

Reply:
I have overhauled the whole kernel specialization code and added new test cases. (`numba_dpex/tests/kernel_tests/test_kernel_specialization`)